A **ML Model** that predicts if your crush likes you or not.  
- Saw this idea on instagram and thought of giving it a try.   
- Fun to learn model training and predicting.
- Lets gooooo!  


---



Library for data handling

In [1]:
import numpy as np
import pandas as pd

Dummy data to train model

In [10]:
training_data = {
    "initiates_conversation": [8, 3, 7, 2, 9, 4, 6, 1, 5, 0, 7, 2, 8, 3, 9, 1, 6, 4, 5, 0, 8, 2, 7, 3, 9, 1, 6, 4, 5, 0],
    "engagement": [9, 4, 8, 3, 7, 2, 6, 1, 5, 0, 8, 3, 7, 2, 9, 1, 6, 4, 5, 0, 9, 2, 8, 3, 7, 1, 6, 4, 5, 0],
    "quick_responses": [9, 1, 5, 0, 8, 3, 7, 4, 6, 2, 8, 2, 6, 1, 9, 3, 7, 4, 5, 0, 9, 1, 6, 2, 7, 3, 8, 4, 5, 0],
    "eye_contact": [7, 3, 8, 1, 9, 5, 6, 4, 7, 2, 8, 2, 7, 3, 9, 1, 6, 4, 5, 0, 9, 3, 7, 2, 8, 1, 6, 4, 5, 0],
    "joke_responses": [8, 2, 6, 1, 7, 3, 9, 5, 4, 0, 7, 3, 8, 2, 9, 1, 6, 4, 5, 0, 8, 2, 7, 3, 9, 1, 6, 4, 5, 0],
    "nervous": [6, 1, 4, 0, 5, 2, 7, 3, 5, 1, 4, 2, 6, 1, 5, 0, 7, 3, 6, 2, 5, 1, 4, 0, 6, 2, 7, 3, 5, 1],
    "stays_near_you": [9, 2, 5, 1, 8, 3, 7, 4, 6, 0, 8, 2, 6, 1, 9, 3, 7, 4, 5, 0, 9, 2, 6, 1, 8, 3, 7, 4, 5, 0],
    "helps_you": [7, 4, 6, 1, 9, 9, 8, 3, 7, 0, 8, 4, 7, 2, 9, 1, 6, 3, 5, 0, 9, 4, 7, 2, 8, 1, 6, 3, 5, 0],
    "smiles": [9, 3, 7, 2, 8, 4, 9, 5, 6, 1, 8, 2, 7, 3, 9, 1, 6, 4, 5, 0, 9, 3, 7, 2, 8, 1, 6, 4, 5, 0],
    "takes_you_out": [8, 1, 5, 0, 7, 3, 9, 2, 6, 1, 7, 2, 6, 1, 9, 3, 8, 4, 5, 0, 9, 2, 7, 3, 8, 1, 6, 4, 5, 0],

    "target": [92, 35, 78, 22, 96, 40, 70, 18, 65, 10, 85, 30, 75, 25, 98, 38, 72, 28, 68, 12, 90, 33, 80, 26, 88, 20, 74, 29, 64, 14]
}


In [11]:
df = pd.DataFrame(training_data)
# print(df.info())

Define features and target

In [12]:
X = df.drop("target", axis=1)
y = df["target"]

In [13]:
X = X.to_numpy()
y = y.to_numpy()

# print(X)
# print(y)

Loading CSV file : in case we have a bigger dataset

In [6]:
# df = pd.read_csv('training_data.csv')

X = df.drop('target', axis = 1).values
y = df['target'].values

Library to use regression model

In [14]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

Model training

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [16]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

Model evaluation

In [17]:
predicted_test = model.predict(X_test)
# print(predicted_test)

MAE calculation

In [19]:
from sklearn.metrics import mean_absolute_error

mae = mean_absolute_error(y_test, predicted_test)
print("Model: Linear Regression")
print(f"Error: {mae:.2f}%  (kind of)")

Model: Linear Regression
Error: 11.17%  (kind of)


R^2 calculation

In [20]:
r2_score = model.score(X_test, y_test)
print("Model: Linear Regression")
print(f"Accuracy: {r2_score*100:.2f}% (kind of)")

Model: Linear Regression
Accuracy: 35.85% (kind of)


The R^2 score is low, exploring ridge regression now

In [21]:
from sklearn.linear_model import Ridge

In [22]:
ridge_model = Ridge(alpha = 1.0)

ridge_model.fit(X_train, y_train)
ridge_pred_test = ridge_model.predict(X_test)
ridge_r2 = ridge_model.score(X_test, y_test)
print("Model: Ridge Regression")
print(f"Accuracy: {ridge_r2*100:.2f}% (kind of)")

Model: Ridge Regression
Accuracy: 35.57% (kind of)


Implementing RidgeCV

In [23]:
alphaa = [0.01, 0.6, 1.0, 1.5, 2.0, 10, 100.0]

from sklearn.linear_model import RidgeCV
ridge_cv = RidgeCV(alphas = alphaa)
ridge_cv.fit(X_train, y_train)
ridgeCV_r2 = ridge_cv.score(X_test, y_test)

print("Model: RidgeCV")
print("Optimal alpha:", ridge_cv.alpha_)
print(f"Accuracy: {ridgeCV_r2*100:.2f}% (kind of)")

Model: RidgeCV
Optimal alpha: 2.0
Accuracy: 35.71% (kind of)


As real love is messy and randomness is handled more effectively by non-linear models, experimenting with `RandomForestRegressor`

In [24]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

X = df.drop("target", axis=1)
y = df["target"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

random_forest = RandomForestRegressor(
    n_estimators=300,
    max_depth=8,
    random_state=42
)

random_forest.fit(X_train, y_train)
pred = random_forest.predict(X_test)
random_forest_r2 = r2_score(y_test, pred)

print("Model: Random Forest Regressor")
print(f"Accuracy: {random_forest_r2*100:.2f}% (kind of)")

Model: Random Forest Regressor
Accuracy: 77.10% (kind of)


Saving the best model : Random Forest Regressor (Accuracy: 77.10%)

In [25]:
import joblib
joblib.dump(random_forest, 'crush_predictor_model.pkl')

['crush_predictor_model.pkl']

User interface

In [26]:
ratings = []
features = [
    "💬 Do they start conversations with you?",
    "🎧 Do they seem interested when you talk?",
    "📱 Do they reply quickly to your messages?",
    "👀 Do they make eye contact often?",
    "😄 Do they joke around with you?",
    "😬 Do they seem nervous around you sometimes?",
    "🧍‍♂️ Do they stay close to you in groups?",
    "🤝 Do they help you when you need it?",
    "😊 Do they smile a lot around you?",
    "🎉 Do they invite you to hang out?"
]

print("Rate your experiences with your crush on below factors (0-10): ")
for f in features:
  rating = int(input(f"{f}: "))
  if(rating>10):
    ratings.append(10)
  else:
    ratings.append(rating)

print(ratings)

Rate your experiences with your crush on below factors (0-10): 
💬 Do they start conversations with you?: 5
🎧 Do they seem interested when you talk?: 6
📱 Do they reply quickly to your messages?: 7
👀 Do they make eye contact often?: 8
😄 Do they joke around with you?: 9
😬 Do they seem nervous around you sometimes?: 10
🧍‍♂️ Do they stay close to you in groups?: 7
🤝 Do they help you when you need it?: 6
😊 Do they smile a lot around you?: 5
🎉 Do they invite you to hang out?: 5
[5, 6, 7, 8, 9, 10, 7, 6, 5, 5]


Reshaping the data : Required when we use linear models.  
Here `RandomForestRegressor` is used so no need to run this block

In [ ]:
# ratings = np.array(ratings)
# user_input = ratings.reshape(1,10)

Prediction process

In [27]:
loaded_model = joblib.load("crush_predictor_model.pkl")
print("Model successfully loaded!")

columns = [
    "initiates_conversation",
    "engagement",
    "quick_responses",
    "eye_contact",
    "joke_responses",
    "nervous",
    "stays_near_you",
    "helps_you",
    "smiles",
    "takes_you_out"
]

user_df = pd.DataFrame([ratings], columns=columns)
# print(user_df)
predicted_value = np.clip(loaded_model.predict(user_df), 0, 200)

print(f"Your crush's likelihood of liking you back is {predicted_value[0]:.2f}%")

Model successfully loaded!
Your crush's likelihood of liking you back is 71.90%
